Fix secret-config-kv-prefix-path property

vsevel opened the pull request:
This changes property `quarkus.vault.secret-config-kv-path.singer=multi/singer1,multi/singer2` into `quarkus.vault.secret-config-kv-prefix.singer.paths=multi/singer1,multi/singer2` (configuration breaking change)
The original approach (reusing the same name for the prefixed properties and non prefixed properties) seemed to work, but only because the re-parsing of configuration done in the `VaultConfigSource` was doing the right thing.

Fixes https://github.com/quarkusio/quarkus/issues/13276

vsevel committed:
Fix secret-config-kv-prefix-path property #13300

sberyozkin commented on integration-tests/vault/src/test/resources/application-vault-multi-path.properties:
```diff
@@ -2,7 +2,7 @@ quarkus.vault.url=https://localhost:8200
 quarkus.vault.authentication.userpass.password=sinclair
 quarkus.vault.secret-config-kv-path=multi/default1,multi/default2
-quarkus.vault.secret-config-kv-path.singer=multi/singer1,multi/singer2
+quarkus.vault.secret-config-kv-prefix.singer.paths=multi/singer1,multi/singer2
```
Hi Vincent @vsevel 

Can you please remind what was the difference between
```
quarkus.vault.secret-config-kv-path=multi/default1,multi/default2
```
and
```
quarkus.vault.secret-config-kv-path.singer=multi/singer1,multi/singer2
```
?
Please also update the docs

vsevel replied:
@sberyozkin hello Sergey,  `quarkus.vault.secret-config-kv-path=multi/default1,multi/default2` will mount the content of paths `multi/default1` and `multi/default2` as MP config properties with no prefix. 
so if `foo=bar` is in path `multi/default1`, you will be able to read `foo` by doing:
```
@ConfigProperty(name = "foo")
String foo;
```

whereas `quarkus.vault.secret-config-kv-prefix.singer.paths=multi/singer1,multi/singer2` will mount those properties with a prefix. so if `bob=dylan` is in `multi/singer1`, then you will be able to access it through:
```
@ConfigProperty(name = "singer.bob")
String bob;
```

this allows to handle name conflicts of properties named identically in different paths. the most obvious example is if multiple paths contain key `password`.

> Please also update the docs

doc is up to date. the only example in the guides was for property `quarkus.vault.secret-config-kv-path`, which did not change. `quarkus.vault.secret-config-kv-prefix.singer.paths` is documented through the javadoc, which is up to date.

cheers.

sberyozkin replied:
Hi Vincent @vsevel Thanks; your explanation about the docs makes sense, and indeed, about the properties, I just wondering if the one extra segment in a new property name can be collapsed.
It was :
```
quarkus.vault.secret-config-kv-path.singer=multi/singer1,multi/singer2
```
now it is
```
quarkus.vault.secret-config-kv-prefix.singer.paths=multi/singer1,multi/singer2
```

So instead of `secret-config-kv-path` it is `secret-config-kv-prefix`, and instead of `singer` (in this concrete case) it is `singer.paths`, 
but we also still have the original non prefix case:
```
quarkus.vault.secret-config-kv-path=multi/default1,multi/default2
```

where it is `secret-config-kv-path` (not `secret-config-kv-paths`).

it feels like it can be simpler (possibly more consistent) to have
```
quarkus.vault.secret-config-kv-path-prefix.singer=multi/singer1,multi/singer2
```

and then we'd have

```
# without prefix
quarkus.vault.secret-config-kv-path=multi/default1,multi/default2

# with prefix, instead of the old
# quarkus.vault.secret-config-kv-path.singer=multi/singer1,multi/singer2
quarkus.vault.secret-config-kv-path-prefix.singer=multi/singer1,multi/singer2
```

What do you think ?

vsevel replied:
sounds good. I remember trying different things. I am wondering if I got some issues with: `Map<String, List<String>>`. I am going to try.

vsevel replied:
so I checked. technically it works. but there is an issue with doc generation.
in the configuration reference property name is `quarkus.vault.secret-config-kv-path-prefix`, and we would want it to be something like `quarkus.vault.secret-config-kv-path-prefix."prefix"`. 
and the type is `Map<String,List<String>>`, when that should be `string` (like in the `quarkus.vault.credentials-provider."credentials-provider".kv-key` property), or better `string list`.
so what should we do:
 - go on with your proposition, and accept the approximate documentation (and may be create an improvement issue)
 - or stay with I was able to make work

sberyozkin replied:
I'm not sure I understand the doc gen issue. @gsmet Hi Guiilaume, can you comment please as well ?

vsevel replied:
to be explicit, here is the generated html with property `public Map<String, List<String>> secretConfigKvPathPrefix;`:

```
<tr>
<td class="tableblock halign-left valign-top"><div class="content"><div class="paragraph">
<p><a id="quarkus-vault_quarkus.vault.secret-config-kv-path-prefix-secret-config-kv-path-prefix"></a><code><a href="#quarkus-vault_quarkus.vault.secret-config-kv-path-prefix-secret-config-kv-path-prefix">quarkus.vault.secret-config-kv-path-prefix</a></code></p>
</div>
<div class="openblock description">
<div class="content">
<div class="paragraph">
<p>List of comma separated vault paths in kv store,
where all properties will be available as <strong>prefixed</strong> MP config properties.</p>
</div>
<div class="paragraph">
<p>For instance if the application properties contains
<code>quarkus.vault.secret-config-kv-path-prefix.myprefix=config</code>, and
vault path <code>secret/config</code> contains <code>foo=bar</code>, then <code>myprefix.foo</code>
will be available in the MP config.</p>
</div>
<div class="paragraph">
<p>If the same property is available in 2 different paths for the same prefix, the last one
will win.</p>
</div>
<div class="paragraph">
<p>See also the documentation for the <code>kv-secret-engine-mount-path</code> property for some insights on how
the full Vault url gets built.</p>
</div>
</div>
</div></div></td>
<td class="tableblock halign-left valign-middle"><p class="tableblock"><code>Map&lt;String,List&lt;String&gt;&gt;</code></p></td>
<td class="tableblock halign-left valign-middle"></td>
</tr>
```

You can see the name is `quarkus.vault.secret-config-kv-path-prefix` and the type is `Map&lt;String,List&lt;String&gt;&gt;`

gsmet replied:
@vsevel can you add a commit with your new proposal so that @machi1990 can have a look to the doc generation issue? Thanks!

machi1990 replied:
@vsevel tag me along once that is done. I'll be able to have a look at this as part of this PR or on separate follow up PR.

vsevel replied:
hi @machi1990 thanks for your help
I have committed the change
here is the associated [code](https://github.com/vsevel/quarkus/blob/vault_map_prop/extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/VaultRuntimeConfig.java#L117-L136)

machi1990 replied:
Thanks @vsevel looking at the thread again,  normally a `@ConfigDocMapKey("prefix")` should have fixed the issue. 
but this won't work since the map is a "pass through map" i.e it does not point to a `@ConfigGroup` but point to a terminal type e.g `String`, `List`, `Duration` etc. For those kind of maps we do not append the prefix but instead show the type of the map with expected generic values in this case `Map<String, List<String>>`. 

I've added https://github.com/vsevel/quarkus/commit/cd98df9d1aad6827e4a3001df531a6540337a384 to the on top of your change. Let me know if that works for you.

vsevel replied:
@machi1990 I have seen your commit, and I think you are reverting back to my original `quarkus.vault.secret-config-kv-prefix.X.paths` proposition that @sberyozkin challenged, and counter-proposed `quarkus.vault.secret-config-kv-path-prefix.X`
I was able to make this work. but it did not feel right to Sergey, and I agree with him on this.

machi1990 replied:
> @machi1990 I have seen your commit, and I think you are reverting back to my [original](https://github.com/quarkusio/quarkus/commit/1eca99cfcb40d131be7ace1872635990cfe734f5#diff-ccdd86f7203916d7b0e179ba75491f1b532fe12f8c747f676b28ef0a05950420R119-R120) `quarkus.vault.secret-config-kv-prefix.X.paths` proposition that @sberyozkin challenged, and counter-proposed `quarkus.vault.secret-config-kv-path-prefix.X`
> I was able to make this work. but it did not feel right to Sergey, and I agree with him on this.

@vsevel not quite, I am using parent naming


```java
  @ConfigItem(name = ConfigItem.PARENT)
   List<String> paths;
```

This means that the `paths`  will inherit the enclosing group name. See a screenshot of the attached doc.

![Screenshot from 2020-11-23 20-46-56](https://user-images.githubusercontent.com/10106536/100008051-38d59880-2dcd-11eb-8a64-78b2cef04368.png)

machi1990 replied:
@vsevel I noticed a test failure and had to force push on top of my initial commit.  

Do you mind re-checking this and see if it fits @sberyozkin for the suggested naming and your own requirement for the doc?

vsevel replied:
hi @machi1990 it looks very good indeed. 2 minor comments: is it possible to make it _not_ required (i.e. the map could be empty)? and would it be possible to make it named `quarkus.vault.secret-config-kv-path."prefix"` instead of `quarkus.vault.secret-config-kv-path-prefix."prefix"` (which was @sberyozkin 's initial [request](https://github.com/quarkusio/quarkus/pull/13300#discussion_r526019797)) - knowing that there is already a property `quarkus.vault.secret-config-kv-path.` (with no _prefix_), which might end up colliding?

machi1990 replied:
I think `Map` are optional by default. In regard to the name "quarkus.vault.secret-config-kv-path-prefix."prefix"", I think if we change the field name of the map, it will do it. I might have missed it when doing the refactor. I'll push a commit.

vsevel replied:
hello @machi1990, I specified:

```
    @ConfigItem(name = "secret-config-kv-path")
    @ConfigDocMapKey("prefix")
    public Map<String, KvPathConfig> secretConfigKvPathPrefix;
```

Then I tried 2 things:

**without `public Optional<List<String>> secretConfigKvPath`**

with properties:
```
quarkus.vault.secret-config-kv-path=multi/default1,multi/default2
quarkus.vault.secret-config-kv-path.singer=multi/singer1,multi/singer2
```

I get:
```
secretConfigKvPathPrefix = {singer=SecretConfigKvPathConfig{paths=[multi/singer1, multi/singer2]}}
```
so this does not work as hoped. we would have expected to have another entry for entry key `""` and value `multi/default1,multi/default2`.

**with `public Optional<List<String>> secretConfigKvPath`**

I get:

```
config secretConfigKvPath = Optional[[multi/default1, multi/default2]]
config secretConfigKvPathPrefix = {singer=SecretConfigKvPathConfig{paths=[multi/singer1, multi/singer2]}}
```

so we have to have the additional `secretConfigKvPath` variable, but that is a very small compromise.

I tried also using an `Optional<Map<String, KvPathConfig>> secretConfigKvPathPrefix`, but initialization fails with a ` No Converter registered for class io.quarkus.vault.runtime.config.VaultRuntimeConfig$KvPathConfig`.

so apart from having to keep an extra var `secretConfigKvPath` to handle the case of the _empty_ prefix, and the `secretConfigKvPathPrefix` that is marked _required_ in doc this looks perfect.
any opportunity to solve those 2 issues, or we call it a day?

thanks again.

machi1990 replied:
Thanks @vsevel for checking it out. 

Let's keep the `public Optional<List<String>> secretConfigKvPath` then and glad the proposed approach work out.

vsevel replied:
sounds good. thanks so much for your help @machi1990

machi1990 commented on extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/VaultSecretConfigKvPrefixPathsConfig.java:
```diff
@@ -0,0 +1,46 @@
+import io.quarkus.runtime.annotations.ConfigItem;
+
+@ConfigGroup
+public class VaultSecretConfigKvPrefixPathsConfig {
```
This class is no longer used, right? We can remove this or and keep https://github.com/quarkusio/quarkus/pull/13300/files#diff-ccdd86f7203916d7b0e179ba75491f1b532fe12f8c747f676b28ef0a05950420R263 with proper renaming or vice versa.

vsevel replied:
you are correct. I have cleaned up the obsolete code. thanks

vsevel commented:
hello all, I have squashed commits and rebased. I think the result is looking good thanks for @machi1990 's help.
@sberyozkin feel free to take a look and make sure its matches your expectations.

sberyozkin commented:
Hi @vsevel, @machi1990  Thanks, it looks good; @machi1990 thanks for your help :-)

sberyozkin reviewed (approved):
@vsevel, Gradle test is failing, most likely it is unrelated, but please rebase one more time

machi1990 commented:
> Hi @vsevel, @machi1990 Thanks, it looks good; @machi1990 thanks for your help :-)

> hello all, I have squashed commits and rebased. I think the result is looking good thanks for @machi1990 's help.
> @sberyozkin feel free to take a look and make sure its matches your expectations.

You are welcome :-)

vsevel commented:
@sberyozkin build failed again, this time on the Microprofile TCK

vsevel commented:
@sberyozkin failed build again because of the global timeout set on the maven build (3h). not too sure what I should do.

sberyozkin commented:
@vsevel IMHO it is save to merge this PR as all the Vault tests pass

vsevel commented:
thanks to all of you for your assistance in the review and figuring out how to implement the best solution
